libc/src/math/generic/asinf16.cpp

```diff
@@ -0,0 +1,135 @@
+//===-- Half-precision asinf16(x) function --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/math/asinf16.h"
+#include "hdr/errno_macros.h"
+#include "hdr/fenv_macros.h"
+#include "src/__support/FPUtil/FEnvImpl.h"
+#include "src/__support/FPUtil/FPBits.h"
+#include "src/__support/FPUtil/PolyEval.h"
+#include "src/__support/FPUtil/cast.h"
+#include "src/__support/FPUtil/multiply_add.h"
+#include "src/__support/FPUtil/sqrt.h"
+#include "src/__support/macros/optimization.h"
+
+namespace LIBC_NAMESPACE_DECL {
+
+// Generated by Sollya using the following command:
+// > round(pi/2, D, RN);
+static constexpr float PI_2 = 0x1.921fb54442d18p0f;
+
+LLVM_LIBC_FUNCTION(float16, asinf16, (float16 x)) {
+  using FPBits = fputil::FPBits<float16>;
+  FPBits xbits(x);
+
+  uint16_t x_u = xbits.uintval();
+  uint16_t x_abs = x_u & 0x7fff;
+  float xf = x;
+
+  float xsq = xf * xf;
+
+  // |x| <= 0x1p-1, |x| <= 0.5
+  if (x_abs <= 0x3800) {
+    // asinf16(NaN) = NaN
+    if (xbits.is_nan()) {
```

Comment:
This still computes `xf * xf` before checking that `x` is not NaN, and `xbits.is_nan()` is always false here since it implies `x_abs > 0x7c00`.

```suggestion
  // |x| <= 0x1p-1, |x| <= 0.5, or x is NaN.
  if (LIBC_UNLIKELY(x_abs <= 0x3800 || x_abs > 0x7c00)) {
    // asinf16(NaN) = NaN
    if (xbits.is_nan()) {
      // ...
    }

    // ...
  }

  // ...
  float xsq = xf * xf;
```

Reply:
> What about directly checking `xbits.is_nan()`, like so?

It might generate more instructions than `x_abs > 0x7c00`, although I haven't verified it in this case, and I haven't benchmarked it.

> Also, why include the `x_abs <= 0x3800` in the NaN condition? If it falls within the domain of the function, it couldn't be NaN, right?

The point is to group all unlikely cases under a single `if (LIBC_UNLIKELY(...))` statement so that the compiler generates better code for the likely case: https://godbolt.org/z/3ern655Ph.

Reply:
How can I quickly verify the instructions generated?

Reply:
`llvm-objdump -Cdr path/to/*.o`

Reply:
```
// |x| > 0x1p0, |x| > 1, or x is NaN.
if (LIBC_UNLIKELY(x_abs > 0x3c00)) {
  // asinf16(NaN) = NaN
  if (xbits.is_nan()) {
    if (xbits.is_signaling_nan()) {
      fputil::raise_except_if_required(FE_INVALID);
      return FPBits::quiet_nan().get_val();
    }

    return x;
  }

  // 1 < |x| <= +/-inf
  fputil::raise_except_if_required(FE_INVALID);
  fputil::set_errno_if_required(EDOM);

  return FPBits::quiet_nan().get_val();
}
```
> It might generate more instructions than x_abs > 0x7c00, although I haven't verified it in this case, and I haven't benchmarked it

`xbits.is_nan()` and `x_abs > 0x7c00` generate the same instructions. Also, what do you think of having the final check for `inf` values in the same conditional scope as the check for NaN-ness like above?